modules/openid_connect/app/services/openid_connect/user_tokens/fetch_service.rb

```diff
@@ -0,0 +1,121 @@
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module OpenIDConnect
+  module UserTokens
+    ##
+    # Provides APIs to obtain access tokens of a given user for use at a third-party
+    # application for which we know the audience name, which is typically the application's
+    # client_id at an identity provider that OpenProject and the application have in common.
+    class FetchService
+      include Dry::Monads[:result]
+
+      def initialize(user:, jwt_parser: JwtParser.new(verify_audience: false, verify_expiration: false))
+        @user = user
+        @provider = user.authentication_provider
+        @jwt_parser = jwt_parser
+      end
+
+      ##
+      # Obtains an access token that can be used to make requests in the user's name at the
+      # remote service identified by the +audience+ parameter.
+      #
+      # The access token will be refreshed before being returned by this method, if it can be
+      # identified as being expired. There is no guarantee that all access tokens will be properly
+      # recognized as expired, so client's still need to make sure to handle rejected access tokens
+      # properly. Also see #refreshed_access_token_for.
+      def access_token_for(audience:)
+        token = token_with_audience(audience)
+        token = token.bind do |t|
+          if expired?(t.access_token)
+            refresh(t)
+          else
+            Success(t)
+          end
+        end
+
+        token.fmap(&:access_token)
+      end
+
+      ##
+      # Obtains an access token that can be used to make requests in the user's name at the
+      # remote service identified by the +audience+ parameter.
+      #
+      # The access token will always be refreshed before being returned by this method.
+      # It is advised to use this method, after learning that a remote service rejected
+      # an access token, because it was expired.
+      def refreshed_access_token_for(audience:)
+        token_with_audience(audience)
+          .bind { |t| refresh(t) }
+          .fmap(&:access_token)
+      end
+
+      private
+
+      def token_with_audience(aud)
+        token = @user.oidc_user_tokens.where("audiences ? :aud", aud:).first
+        token ? Success(token) : Failure("No token for given audience")
+      end
+
+      def refresh(token)
+        refresh_token_request(token.refresh_token).bind do |json|
```

Comment:
:red_circle: We should handle the case of having no `refresh_token` available, i.e. refresh being impossible.

For the scope of this PR, that's simply an error. For later scope, this might as well simply result in an attempt to exchange a fresh token.

Reply:
done